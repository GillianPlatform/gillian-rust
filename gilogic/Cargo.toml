[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }