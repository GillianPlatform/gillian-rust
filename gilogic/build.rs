fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
