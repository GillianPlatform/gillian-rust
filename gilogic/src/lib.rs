#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
