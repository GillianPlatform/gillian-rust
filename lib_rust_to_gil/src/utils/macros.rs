macro_rules! fatal {
  ($e: expr, $($tts:tt)*) => {
      $e.ty_ctxt.sess.fatal(&format!($($tts)*))
  };
}

pub(crate) use fatal;
