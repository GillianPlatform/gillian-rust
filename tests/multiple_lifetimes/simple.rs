//@check-pass
extern crate gilogic;
use gilogic::macros::*;
use gilogic::Ownable;

// #[show_safety]
fn swap_borrows<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> (&'b mut i32, &'a mut i32) {
  (y, x)
}

#[show_safety]
fn fst<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> (&'a mut i32) {
  x
}


#[show_safety]
fn snd<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> (&'b mut i32) {
  y
}

// #[show_safety]
// fn swap<T: Ownable, U : Ownable>(x : T, y : U) -> (U, T) {
//   (y, x)
// }

// #[show_safety]
// fn swap_borrows2<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> (&'b mut i32, &'a mut i32) {
//   swap(x, y)
// }

// fn test() {
//   let mut x = 0;
//   let mut y = 1;

//   let (a, b) = swap_borrows2(&mut x, &mut y);
//   let (c, d) = swap_borrows(a, b);

//   *c = 5;
//   *d = 10;

//   assert!(x == 5);
//   assert!(y == 10);
// }