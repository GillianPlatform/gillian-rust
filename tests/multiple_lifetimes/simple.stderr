[TIMESTAMP WARN  lib_rtg::metadata] could not read metadata for crate `"std"`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
DefId(0:9 ~ simple[d6ea]::fst_spec_15d57891_48b9_42ec_9cc2_f95df287dab8)
warning: unused variable: `y`
  --> ../tests/multiple_lifetimes/simple.rs:13:33
   |
13 | fn fst<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> &'a mut i32 {
   |                                 ^ help: if this is intentional, prefix it with an underscore: `_y`
   |
   = note: `#[warn(unused_variables)]` on by default

DefId(0:17 ~ simple[d6ea]::snd_spec_9d539363_def2_4fce_a949_89417687c2e0)
warning: unused variable: `x`
  --> ../tests/multiple_lifetimes/simple.rs:19:16
   |
19 | fn snd<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> &'b mut i32 {
   |                ^ help: if this is intentional, prefix it with an underscore: `_x`

compiling DefId(20:155 ~ gilogic[8049]::ownable::{impl#0}::own)
DefId(20:155 ~ gilogic[8049]::ownable::{impl#0}::own)
warning: function `swap_borrows` is never used
 --> ../tests/multiple_lifetimes/simple.rs:8:4
  |
8 | fn swap_borrows<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> (&'b mut i32, &'a mut i32) {
  |    ^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: function `fst` is never used
  --> ../tests/multiple_lifetimes/simple.rs:13:4
   |
13 | fn fst<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> &'a mut i32 {
   |    ^^^

warning: function `snd` is never used
  --> ../tests/multiple_lifetimes/simple.rs:19:4
   |
19 | fn snd<'a, 'b>(x : &'a mut i32, y: &'b mut i32) -> &'b mut i32 {
   |    ^^^

warning: 5 warnings emitted

[TIMESTAMP DEBUG rust_to_gil] Correct!
